README.rst
Flask-Injector
==============

.. image:: https://codecov.io/gh/alecthomas/flask_injector/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/alecthomas/flask_injector

.. image:: https://github.com/alecthomas/flask_injector/workflows/CI/badge.svg
   :target: https://github.com/alecthomas/flask_injector?query=workflow%3ACI+branch%3Amaster


Adds `Injector <https://github.com/alecthomas/injector>`_ support to Flask,
this way there's no need to use global Flask objects, which makes testing simpler.

Injector is a dependency-injection framework for Python, inspired by Guice. You
can `find Injector on PyPI <https://pypi.org/project/injector/>`_ and `Injector
documentation on Read the Docs <https://injector.readthedocs.io/en/latest/>`_.

`Flask-Injector` is compatible with CPython 3.7+.
As of version 0.15.0 it requires Injector version 0.20.0 or greater and Flask
2.2.0 or greater.

GitHub project page: https://github.com/alecthomas/flask_injector

PyPI package page: https://pypi.org/project/Flask-Injector/

Changelog: https://github.com/alecthomas/flask_injector/blob/master/CHANGELOG.rst

Features
--------

Flask-Injector lets you inject dependencies into:

* views (functions and class-based)
* `before_request` handlers
* `after_request` handlers
* `teardown_request` handlers
* template context processors
* error handlers
* Jinja environment globals (functions in `app.jinja_env.globals`)
* Flask-RESTFul Resource constructors
* Flask-RestPlus Resource constructors
* Flask-RESTX Resource constructors

Flask-Injector supports defining types using function annotations (Python 3),
see below.

Documentation
-------------

As Flask-Injector uses Injector under the hood you should find the
`Injector documentation <https://injector.readthedocs.io/en/latest/>`_,
including the `Injector API reference <https://injector.readthedocs.io/en/latest/api.html>`_,
helpful. The `Injector README <https://github.com/alecthomas/injector/blob/master/README.md>`_
provides a tutorial-level introduction to using Injector.

The Flask-Injector public API consists of the following:

* `FlaskInjector` class with the constructor taking the following parameters:

  * `app`, an instance of`flask.Flask` [mandatory] – the Flask application to be used
  * `modules`, an iterable of
    `Injector modules <https://injector.readthedocs.io/en/latest/api.html#injector.Binder.install>`_ [optional]
    – the Injector modules to be used.
  * `injector`, an instance of
    `injector.Injector <https://injector.readthedocs.io/en/latest/api.html#injector.Injector>`_ [optional]
    – an instance of Injector to be used if, for some reason, it's not desirable
    for `FlaskInjector` to create a new one. You're likely to not need to use this.
  * `request_scope_class`, an `injector.Scope <https://injector.readthedocs.io/en/latest/api.html#injector.Scope>`_
    subclass [optional] – the scope to be used instead of `RequestScope`. You're likely to need to use this
    except for testing.
* `RequestScope` class – an `injector.Scope <https://injector.readthedocs.io/en/latest/api.html#injector.Scope>`_
  subclass to be used for storing and reusing request-scoped dependencies
* `request` object – to be used as a class decorator or in explicit
  `bind() <https://injector.readthedocs.io/en/latest/api.html#injector.Binder.bind>`_ calls in
  Injector modules.
  
Creating an instance of `FlaskInjector` performs side-effectful configuration of the Flask
application passed to it. The following bindings are applied (if you want to modify them you
need to do it in one of the modules passed to the `FlaskInjector` constructor):

* `flask.Flask` is bound to the Flask application in the (scope: singleton)
* `flask.Config` is bound to the configuration of the Flask application
* `flask.Request` is bound to the current Flask request object, equivalent to the thread-local
  `flask.request` object (scope: request)
 
Example application using Flask-Injector
----------------------------------------

.. code:: python

    import sqlite3
    from flask import Flask, Config
    from flask.views import View
    from flask_injector import FlaskInjector
    from injector import inject

    app = Flask(__name__)

    # Configure your application by attaching views, handlers, context processors etc.:

    @app.route("/bar")
    def bar():
        return render("bar.html")


    # Route with injection
    @app.route("/foo")
    def foo(db: sqlite3.Connection):
        users = db.execute('SELECT * FROM users').all()
        return render("foo.html")


    # Class-based view with injected constructor
    class Waz(View):
        @inject
        def __init__(self, db: sqlite3.Connection):
            self.db = db

        def dispatch_request(self, key):
            users = self.db.execute('SELECT * FROM users WHERE name=?', (key,)).all()
            return 'waz'

    app.add_url_rule('/waz/<key>', view_func=Waz.as_view('waz'))


    # In the Injector world, all dependency configuration and initialization is
    # performed in modules (https://injector.readthedocs.io/en/latest/terminology.html#module).
    # The same is true with Flask-Injector. You can see some examples of configuring
    # Flask extensions through modules below.

    # Accordingly, the next step is to create modules for any objects we want made
    # available to the application. Note that in this example we also use the
    # Injector to gain access to the `flask.Config`:

    def configure(binder):
        binder.bind(
            sqlite3.Connection,
            to=sqlite3.Connection(':memory:'),
            scope=request,
        )
    
    # Initialize Flask-Injector. This needs to be run *after* you attached all
    # views, handlers, context processors and template globals.

    FlaskInjector(app=app, modules=[configure])

    # All that remains is to run the application

    app.run()

See `example.py` for a more complete example, including `Flask-SQLAlchemy` and
`Flask-Cache` integration.

Supporting Flask Extensions
---------------------------

Typically, Flask extensions are initialized at the global scope using a
pattern similar to the following.

.. code:: python

    app = Flask(__name__)
    ext = ExtClass(app)

    @app.route(...)
    def view():
        # Use ext object here...

As we don't have these globals with Flask-Injector we have to configure the
extension the Injector way - through modules. Modules can either be subclasses
of `injector.Module` or a callable taking an `injector.Binder` instance.

.. code:: python

    from injector import Module

    class MyModule(Module):
        @provider
        @singleton
        def provide_ext(self, app: Flask) -> ExtClass:
            return ExtClass(app)

    def main():
        app = Flask(__name__)
        app.config.update(
            EXT_CONFIG_VAR='some_value',
        )

        # attach your views etc. here

        FlaskInjector(app=app, modules=[MyModule])

        app.run()

*Make sure to bind extension objects as singletons.*


CHANGELOG.rst
Flask-Injector Changelog
========================

Version 0.15.0
--------------

* Fixed Flask 2.2.0+ compatibility, thanks to Valentin Baert

Backwards incompatible:

* Dropped support for Flask older than 2.2.0

Version 0.14.0
--------------

* Infrastructure contributions thanks to ZHANG Cheng
* Fixed all known Flask/Werkzeug compatibility issues

Backwards incompatible:

* Dropped Python 3.6 support
* Flask >= 2.1.2 is now required
* Injector >= 0.20.0 is now required

Version 0.13.0
--------------

* Improved Werkzeug/Flask 2.0 compatibility (fixed the "RuntimeError: Working outside of request context" error)

Backwards incompatible:

* Dropped Python 3.5 support
* Dropped flask_restplus support (the library hasn't been usable with Werkzeug 1.0+)

Version 0.12.3
--------------

* Fixed injecting into blueprint-associated teardown_request handlers

Version 0.12.2
--------------

* Added Flask-RESTX integration (#48, thanks to Michael Bukachi)

Version 0.12.1
--------------

* Stopped unnecessarily installing typing on recent Python versions (thanks to Louis Trezzini)
* Fixed injecting request-scoped dependencies into teradown_request handlers (fix suggested
  by Nick Krichevsky)
* Added PEP 561 py.typed marker so that tools know to use type hints in the package's source

Version 0.12.0
--------------

* Added support for adding instance methods as handlers (#35, thanks to Rene Hollander)

Backwards incompatible:

* Dropped Python 3.4 support
* Dropped Flask < 1.0 support
* Dropped Injector < 0.13.2 support

Version 0.11.0
--------------

* flask_restful is no longer required to be installed when flask_restplus is
  used (#24)
* Added support for injecting into before_first_request functions (#26)

Backwards incompatible:

* Dropped Python 3.3 support

Version 0.10.1
--------------

* Got rid of a deprecation warning when Injector 0.13.2 is used

Version 0.10.0
--------------

* Dropped support for Injector < 0.12 and Flask < 0.12
* Dropped use_annotations constructor parameter (this also fixed compatibility
  with Injector 0.13)
* At least for the time being class-based views' constructors need to be marked
  with @inject in order for dependencies to be injected

Version 0.9.0
-------------

* Fixed a bug that would cause a crash when an nonintrospectable callable
  was registered as, for example, before_request hook
* Fixed support for forward references in type hints
* Added type hints to the codebase

Backwards incompatible:

* Dropped support for Injector 0.10

Version 0.8.0
-------------

* Dropped support for Flask < 0.11, Injector < 0.10, Python 2, PyPy and PyPy 3
  (PyPy 3 will be supported in the future)
* Fixed compatibility with Injector 0.11

Version 0.7.1
-------------

* Fixed Flask 0.11 compatibility, thanks to Philip Jones for the initial patch

Version 0.7.0
-------------

* Added support for injecting into Flask-RestPlus Resource constructors
* Added support for dependencies declared using Python 3-style annotations

Version 0.6.2
-------------

* Fixed a regression introduced in 0.6.1 (requesting an interface bound in
  RequestScope outside request context got broken and would raise
  "AttributeError: scope" exception. Even though it's not a documented
  behaviour it's restored now so that backwards compatibility is preserved.

Version 0.6.1
-------------

* Python 2.6 support dropped
* Fixed a memory leak bug (a reference to thread-identity object would be kept
  forever after a request would be served by particular thread; without greenlet
  package installed thread ids (numeric values) are used so the internal
  dictionary of thread local storage grows forever; when greenlet package is
  installed greenlet objects are used as thread identities by Werkzeug so on top
  of the internal storage growing infinitely all objects referenced by those
  greenlet objects are kept alive; keywords: Eventlet, Gevent, GreenThread). See
  GH issue #9 and pull request #11, thanks to Zi Li for the fix

Version 0.6.0
-------------

* Added support for injecting into Flask-RESTFul Resource constructors

Version 0.5.0
-------------

* Removed ``init_app`` and ``post_init_app`` functions
* Fixed a bug with Flask-Injector modifying possibly shared view generated by View.as_view
  (see GH issue #6, test case provided by Nicholas Hollett)
* Work only with Injector >= 0.9.0 now

Version 0.4.0
-------------

* Deprecated ``init_app`` and ``post_init_app`` in favour of ``FlaskInjector``
* Made Flask error handlers support injection

Version 0.3.4
-------------

* Made it possible to inject into Jinja template globals

Version 0.3.3
-------------

* Accomodated to Injector >= 0.9.0


tests.py
from injector import inject
from flask import Blueprint, Flask

from flask_injector import FlaskInjector, route


def test_injection_in_preconfigured_views():
    pages = Blueprint('pages', __name__)

    @pages.route('/')
    @inject(content=str)
    def index(content):
        return content

    app = Flask(__name__)
    app.register_blueprint(pages)
    app.debug = True

    def conf(binder):
        binder.bind(str, to="something")

    flask_injector = FlaskInjector(modules=[conf], inject_native_views=True)
    flask_injector.init_app(app)

    with app.test_client() as c:
        response = c.get('/')
        assert (response.data == "something"), response.data


def test_multiple_routes():
    @route('/', defaults={'page': 'none'})
    @route('/<page>')
    def index(page=None):
        return page

    app = Flask(__name__)
    flask_injector = FlaskInjector([index])
    flask_injector.init_app(app)

    with app.test_client() as c:
        response = c.get('/')
        assert response.data == 'none', response.data
        response = c.get('/boom')
        assert response.data == 'boom', response.data


flask_injector/tests.py
import gc
import json
import warnings
from functools import partial
from typing import NewType

import flask_restful
import flask_restx
from eventlet import greenthread
from injector import __version__ as injector_version, CallableProvider, inject, Scope
from flask import Blueprint, Flask
from flask.templating import render_template_string
from flask.views import View

from flask_injector import request, FlaskInjector


def test_injections():
    l = [1, 2, 3]
    counter = [0]

    def inc():
        counter[0] += 1

    def conf(binder):
        binder.bind(str, to="something")
        binder.bind(list, to=l)

    app = Flask(__name__)

    @app.route('/view1')
    def view1(content: str):
        inc()
        return render_template_string(content)

    class View2(View):
        @inject
        def __init__(self, *args, content: list, **kwargs):
            self.content = content
            super().__init__(*args, **kwargs)

        def dispatch_request(self):
            inc()
            return render_template_string('%s' % self.content)

    @app.before_request
    def br(c: list):
        inc()
        assert c == l

    @app.after_request
    def ar(response_class, c: list):
        inc()
        assert c == l
        return response_class

    @app.context_processor
    def cp(c: list):
        inc()
        assert c == l
        return {}

    @app.teardown_request
    def tr(sender, exc=None, c: list = None):
        inc()
        assert c == l

    app.add_url_rule('/view2', view_func=View2.as_view('view2'))

    FlaskInjector(app=app, modules=[conf])

    client = app.test_client()
    response = client.get('/view1')
    assert response.get_data(as_text=True) == "something"

    response = client.get('/view2')
    assert response.get_data(as_text=True) == '%s' % (l,)

    assert counter[0] == 10


def test_resets():
    app = Flask(__name__)

    counter = [0]

    class OurScope(Scope):
        def __init__(self, injector):
            pass

        def prepare(self):
            pass

        def cleanup(self):
            counter[0] += 1

    @app.route('/')
    def index():
        return 'asd'

    FlaskInjector(app, request_scope_class=OurScope)

    assert counter[0] == 0

    client = app.test_client()
    client.get('/')

    assert counter[0] == 1


def test_memory_leak():
    # The RequestScope holds references to GreenThread objects which would
    # cause memory leak

    # More explanation below
    #
    # In Werkzeug locals are indexed using values returned by ``get_ident`` function:
    #
    # try:
    #    from greenlet import getcurrent as get_ident
    # except ImportError:
    #     try:
    #         from thread import get_ident
    #     except ImportError:
    #         from _thread import get_ident
    #
    # This is what LocalManager.cleanup runs indirectly (__ident_func__
    # points to get_ident unless it's overridden):
    #
    # self.__storage__.pop(self.__ident_func__(), None)

    # If something's assigned in local storage *after* the cleanup is done an entry
    # in internal storage under "the return value of get_ident()" key is recreated
    # and a reference to the key will be kept forever.
    #
    # This is not strictly related to Eventlet/GreenThreads but that's how
    # the issue manifested itself so the test reflects that.
    app = Flask(__name__)

    FlaskInjector(app)

    @app.route('/')
    def index():
        return 'test'

    def get_request():
        with app.test_client() as c:
            c.get('/')

    green_thread = greenthread.spawn(get_request)
    green_thread.wait()
    # Delete green_thread so the GreenThread object is dereferenced
    del green_thread
    # Force run garbage collect to make sure GreenThread object is collected if
    # there is no memory leak
    gc.collect()
    greenthread_count = len([obj for obj in gc.get_objects() if type(obj) is greenthread.GreenThread])

    assert greenthread_count == 0


def test_doesnt_raise_deprecation_warning():
    app = Flask(__name__)

    def provide_str():
        return 'this is string'

    def configure(binder):
        binder.bind(str, to=CallableProvider(provide_str), scope=request)

    @app.route('/')
    def index(s: str):
        return s

    FlaskInjector(app=app, modules=[configure])

    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        with app.test_client() as c:
            c.get('/')
        assert len(w) == 0, map(str, w)


def test_jinja_env_globals_support_injection():
    app = Flask(__name__)

    def configure(binder):
        binder.bind(str, to='xyz')

    def do_something_helper(s: str):
        return s

    app.jinja_env.globals['do_something'] = do_something_helper

    @app.route('/')
    def index():
        return render_template_string('{{ do_something() }}')

    FlaskInjector(app=app, modules=[configure])

    with app.test_client() as c:
        assert c.get('/').get_data(as_text=True) == 'xyz'


def test_error_handlers_support_injection():
    app = Flask(__name__)

    class CustomException(Exception):
        pass

    @app.route('/custom-exception')
    def custom_exception():
        raise CustomException()

    @app.errorhandler(404)
    def handle_404(error, s: str):
        return s, 404

    @app.errorhandler(CustomException)
    def handle_custom_exception(error, s: str):
        return s, 500

    def configure(binder):
        binder.bind(str, to='injected content')

    FlaskInjector(app=app, modules=[configure])

    with app.test_client() as c:
        response = c.get('/this-page-does-not-exist')
        assert (response.status_code, response.get_data(as_text=True)) == (404, 'injected content')

        response = c.get('/custom-exception')
        assert (response.status_code, response.get_data(as_text=True)) == (500, 'injected content')


def test_view_functions_arent_modified_globally():
    # Connected to GH #6 "Doing multiple requests on a flask test client on an injected route
    # fails for all but the first request"
    # The code would modify view functions generated by View.as_view(), it wasn't an issue with
    # views added directly to an application but if function was added to a blueprint and
    # that blueprint was used in multiple applications it'd raise an error

    class MyView(View):
        pass

    blueprint = Blueprint('test', __name__)
    blueprint.add_url_rule('/', view_func=MyView.as_view('view'))

    app = Flask(__name__)
    app.register_blueprint(blueprint)
    FlaskInjector(app=app)

    app2 = Flask(__name__)
    app2.register_blueprint(blueprint)

    # it'd fail here
    FlaskInjector(app=app2)


def test_view_args_and_class_args_are_passed_to_class_based_views():
    class MyView(View):
        def __init__(self, class_arg):
            self.class_arg = class_arg

        def dispatch_request(self, dispatch_arg):
            return '%s %s' % (self.class_arg, dispatch_arg)

    app = Flask(__name__)
    app.add_url_rule('/<dispatch_arg>', view_func=MyView.as_view('view', class_arg='aaa'))

    FlaskInjector(app=app)

    client = app.test_client()
    response = client.get('/bbb')
    print(response.data)
    assert response.data == b'aaa bbb'


def test_flask_restful_integration_works():
    class HelloWorld(flask_restful.Resource):
        @inject
        def __init__(self, *args, int: int, **kwargs):
            self._int = int
            super().__init__(*args, **kwargs)

        def get(self):
            return {'int': self._int}

    app = Flask(__name__)
    api = flask_restful.Api(app)

    api.add_resource(HelloWorld, '/')

    FlaskInjector(app=app)

    client = app.test_client()
    response = client.get('/')
    data = json.loads(response.data.decode('utf-8'))
    assert data == {'int': 0}


def test_flask_restx_integration_works():
    app = Flask(__name__)
    api = flask_restx.Api(app)

    class HelloWorld(flask_restx.Resource):
        @inject
        def __init__(self, *args, int: int, **kwargs):
            self._int = int
            super().__init__(*args, **kwargs)

        @api.doc()
        def get(self):
            return {'int': self._int}

    api.add_resource(HelloWorld, '/hello')

    FlaskInjector(app=app)

    client = app.test_client()
    response = client.get('/hello')
    data = json.loads(response.data.decode('utf-8'))
    assert data == {'int': 0}


def test_request_scope_not_started_before_any_request_made_regression():
    # Version 0.6.1 (patch cacaef6 specifially) broke backwards compatibility in
    # a relatively subtle way. The code used to support RequestScope even in
    # the thread that originally created the Injector object. After cacaef6 an
    # "AttributeError: scope" exception would be raised.
    #
    # For compatibility reason I'll restore the old behaviour, we can
    # deprecate it later if needed

    def configure(binder):
        binder.bind(str, to='this is string', scope=request)

    app = Flask(__name__)
    flask_injector = FlaskInjector(app=app, modules=[configure])
    assert flask_injector.injector.get(str) == 'this is string'


def test_noninstrospectable_hooks_dont_crash_everything():
    app = Flask(__name__)

    def do_nothing():
        pass

    app.before_request(partial(do_nothing))

    # It'd crash here
    FlaskInjector(app=app)


def test_instance_methods():
    class HelloWorldService:
        def get_value(self):
            return "test message 1"

    class HelloWorld:
        def from_injected_service(self, service: HelloWorldService):
            return service.get_value()

        def static_value(self):
            return "test message 2"

    app = Flask(__name__)
    hello_world = HelloWorld()
    app.add_url_rule('/from_injected_service', 'from_injected_service', hello_world.from_injected_service)
    app.add_url_rule('/static_value', 'static_value', hello_world.static_value)

    FlaskInjector(app=app)

    client = app.test_client()
    response = client.get('/from_injected_service')
    assert response.data.decode('utf-8') == "test message 1"

    response = client.get('/static_value')
    assert response.data.decode('utf-8') == "test message 2"


if injector_version >= '0.12':

    def test_forward_references_work():
        app = Flask(__name__)

        @app.route('/')
        def index(x: 'X'):
            return x.message

        FlaskInjector(app=app)

        # The class needs to be module-global in order for the string -> object
        # resolution mechanism to work. I could make it work with locals but it
        # doesn't seem worth it.
        global X

        class X:
            def __init__(self) -> None:
                self.message = 'Hello World'

        try:
            client = app.test_client()
            response = client.get('/')
            assert response.data.decode() == 'Hello World'
        finally:
            del X


def test_request_scope_covers_teardown_request_handlers():
    app = Flask(__name__)
    UserID = NewType('UserID', int)

    @app.route('/')
    def index():
        return 'hello'

    @app.teardown_request
    def on_teardown(exc, user_id: UserID):
        assert user_id == 321

    def configure(binder):
        binder.bind(UserID, to=321, scope=request)

    FlaskInjector(app=app, modules=[configure])
    client = app.test_client()
    response = client.get('/')
    assert response.data.decode() == 'hello'


def test_request_scope_covers_blueprint_teardown_request_handlers():
    app = Flask(__name__)
    UserID = NewType('UserID', int)
    blueprint = Blueprint('blueprint', __name__)

    @blueprint.route('/')
    def index():
        return 'hello'

    @blueprint.teardown_request
    def on_teardown(exc, user_id: UserID):
        assert user_id == 321

    def configure(binder):
        binder.bind(UserID, to=321, scope=request)

    app.register_blueprint(blueprint)
    FlaskInjector(app=app, modules=[configure])
    client = app.test_client()
    response = client.get('/')
    assert response.data.decode() == 'hello'


example.py
# -*- coding: utf-8 -*-
import logging

from injector import Module, Injector, inject, singleton
from flask import Flask, Request, jsonify
from flask_injector import FlaskInjector
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm.exc import NoResultFound
from sqlalchemy import Column, String

il = logging.getLogger('injector')
il.addHandler(logging.StreamHandler())
il.level = logging.DEBUG

"""
This is an example of using Injector (https://github.com/alecthomas/injector) and Flask.

Flask provides a lot of very nice features, but also requires a lot of globals
and tightly bound code. Flask-Injector seeks to remedy this.
"""


# We use standard SQLAlchemy models rather than the Flask-SQLAlchemy magic, as
# it requires a global Flask app object and SQLAlchemy db object.
Base = declarative_base()


class KeyValue(Base):
    __tablename__ = 'data'

    key = Column(String, primary_key=True)
    value = Column(String)

    def __init__(self, key, value):
        self.key = key
        self.value = value

    def serializable(self):
        return


def configure_views(app):
    @app.route('/<key>')
    def get(key, db: SQLAlchemy):
        try:
            kv = db.session.query(KeyValue).filter(KeyValue.key == key).one()
        except NoResultFound:
            response = jsonify(status='No such key', context=key)
            response.status = '404 Not Found'
            return response
        return jsonify(key=kv.key, value=kv.value)

    @app.route('/')
    def list(db: SQLAlchemy):
        data = [i.key for i in db.session.query(KeyValue).order_by(KeyValue.key)]
        return jsonify(keys=data)

    @app.route('/', methods=['POST'])
    def create(request: Request, db: SQLAlchemy):
        kv = KeyValue(request.form['key'], request.form['value'])
        db.session.add(kv)
        db.session.commit()
        response = jsonify(status='OK')
        response.status = '201 CREATED'
        return response

    @app.route('/<key>', methods=['DELETE'])
    def delete(db: SQLAlchemy, key):
        db.session.query(KeyValue).filter(KeyValue.key == key).delete()
        db.session.commit()
        response = jsonify(status='OK')
        response.status = '200 OK'
        return response


class AppModule(Module):
    def __init__(self, app):
        self.app = app

    """Configure the application."""

    def configure(self, binder):
        # We configure the DB here, explicitly, as Flask-SQLAlchemy requires
        # the DB to be configured before request handlers are called.
        db = self.configure_db(self.app)
        binder.bind(SQLAlchemy, to=db, scope=singleton)

    def configure_db(self, app):
        db = SQLAlchemy(app)
        Base.metadata.create_all(db.engine)
        db.session.add_all([KeyValue('hello', 'world'), KeyValue('goodbye', 'cruel world')])
        db.session.commit()
        return db


def main():
    app = Flask(__name__)
    app.config.update(DB_CONNECTION_STRING=':memory:', SQLALCHEMY_DATABASE_URI='sqlite://')
    app.debug = True

    # An explicit context because AppModule calls some flask-sqlalchemy's code that needs
    # a context.
    #
    # https://github.com/pallets-eco/flask-sqlalchemy/issues/1129#issuecomment-1283219340
    with app.app_context():
        injector = Injector([AppModule(app)])
    configure_views(app=app)

    FlaskInjector(app=app, injector=injector)

    client = app.test_client()

    response = client.get('/')
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.post('/', data={'key': 'foo', 'value': 'bar'})
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.get('/')
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.get('/hello')
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.delete('/hello')
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.get('/')
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.get('/hello')
    print('%s\n%s%s' % (response.status, response.headers, response.data))
    response = client.delete('/hello')
    print('%s\n%s%s' % (response.status, response.headers, response.data))


if __name__ == '__main__':
    main()


flask_injector_tests_py3.py
import json
import warnings

import flask_restful
import flask_restplus
from injector import CallableProvider
from flask import Flask
from flask.templating import render_template_string
from flask.views import View
from nose.tools import eq_

from flask_injector import request, FlaskInjector


def test_injections():
    l = [1, 2, 3]
    counter = [0]

    def inc():
        counter[0] += 1

    def conf(binder):
        binder.bind(str, to="something")
        binder.bind(list, to=l)

    app = Flask(__name__)

    @app.route('/view1')
    def view1(content: str):
        inc()
        return render_template_string(content)

    class View2(View):
        def __init__(self, *args, content: list, **kwargs):
            self.content = content
            super().__init__(*args, **kwargs)

        def dispatch_request(self):
            inc()
            return render_template_string('%s' % self.content)

    @app.before_request
    def br(c: list):
        inc()
        eq_(c, l)

    @app.after_request
    def ar(response_class, c: list):
        inc()
        eq_(c, l)
        return response_class

    @app.context_processor
    def cp(c: list):
        inc()
        eq_(c, l)
        return {}

    @app.teardown_request
    def tr(sender, exc=None, c: list = None):
        inc()
        eq_(c, l)

    app.add_url_rule('/view2', view_func=View2.as_view('view2'))

    FlaskInjector(app=app, modules=[conf], use_annotations=True)

    with app.test_client() as c:
        response = c.get('/view1')
        eq_(response.get_data(as_text=True), "something")

    with app.test_client() as c:
        response = c.get('/view2')
        eq_(response.get_data(as_text=True), '%s' % (l,))

    eq_(counter[0], 10)


def test_doesnt_raise_deprecation_warning():
    app = Flask(__name__)

    def provide_str():
        return 'this is string'

    def configure(binder):
        binder.bind(str, to=CallableProvider(provide_str), scope=request)

    @app.route('/')
    def index(s: str):
        return s

    FlaskInjector(app=app, modules=[configure], use_annotations=True)

    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        with app.test_client() as c:
            c.get('/')
        eq_(len(w), 0, map(str, w))


def test_jinja_env_globals_support_injection():
    app = Flask(__name__)

    def configure(binder):
        binder.bind(str, to='xyz')

    def do_something_helper(s: str):
        return s

    app.jinja_env.globals['do_something'] = do_something_helper

    @app.route('/')
    def index():
        return render_template_string('{{ do_something() }}')

    FlaskInjector(app=app, modules=[configure], use_annotations=True)

    with app.test_client() as c:
        eq_(c.get('/').get_data(as_text=True), 'xyz')


def test_error_handlers_support_injection():
    app = Flask(__name__)

    class CustomException(Exception):
        pass

    @app.route('/custom-exception')
    def custom_exception():
        raise CustomException()

    @app.errorhandler(404)
    def handle_404(error, s: str):
        return s, 404

    @app.errorhandler(CustomException)
    def handle_custom_exception(error, s: str):
        return s, 500

    def configure(binder):
        binder.bind(str, to='injected content')

    FlaskInjector(app=app, modules=[configure], use_annotations=True)

    with app.test_client() as c:
        response = c.get('/this-page-does-not-exist')
        eq_((response.status_code, response.get_data(as_text=True)),
            (404, 'injected content'))

        response = c.get('/custom-exception')
        eq_((response.status_code, response.get_data(as_text=True)),
            (500, 'injected content'))


def test_flask_restful_integration_works():

    class HelloWorld(flask_restful.Resource):
        def __init__(self, *args, int: int, **kwargs):
            self._int = int
            super().__init__(*args, **kwargs)

        def get(self):
            return {'int': self._int}

    app = Flask(__name__)
    api = flask_restful.Api(app)

    api.add_resource(HelloWorld, '/')

    FlaskInjector(app=app, use_annotations=True)

    client = app.test_client()
    response = client.get('/')
    data = json.loads(response.data.decode('utf-8'))
    eq_(data, {'int': 0})


def test_flask_restplus_integration_works():
    class HelloWorld(flask_restplus.Resource):
        def __init__(self, *args, int: int, **kwargs):
            self._int = int
            super().__init__(*args, **kwargs)

        def get(self):
            return {'int': self._int}

    app = Flask(__name__)
    api = flask_restplus.Api(app)

    api.add_resource(HelloWorld, '/hello')

    FlaskInjector(app=app, use_annotations=True)

    client = app.test_client()
    response = client.get('/hello')
    data = json.loads(response.data.decode('utf-8'))
    eq_(data, {'int': 0})
